mm2src/coins/sia/src/http_client.rs

```diff
@@ -0,0 +1,168 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SiaHttpConf {
+    pub url: Url,
+    pub password: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct SiaApiClient {
+    client: Client,
+    conf: SiaHttpConf,
+}
+
+// this is neccesary to show the URL in error messages returned to the user
+// this can be removed in favor of using ".with_url()" once reqwest is updated to v0.11.23
+#[derive(Debug)]
+pub struct ReqwestErrorWithUrl {
+    error: ReqwestError,
+    url: Url,
+}
+
+impl Display for ReqwestErrorWithUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error: {}, URL: {}", self.error, self.url)
+    }
+}
+
+// TODO clean up reqwest errors
+// update reqwest to latest for `.with_url()` method
+#[derive(Debug, Display)]
+pub enum SiaApiClientError {
+    Timeout(String),
+    BuildError(String),
+    ServerUnreachable(String),
+    ReqwestFetchError(ReqwestErrorWithUrl), // TODO make an enum
+    ReqwestError(reqwest::Error),
+    ReqwestParseInvalidEncodingError(String),
+    ReqwestParseInvalidJsonError(String),
+    ReqwestParseUnexpectedTypeError(String),
+    ReqwestTlsError(ReqwestErrorWithUrl),
+    UrlParse(url::ParseError),
+    UnexpectedHttpStatus(u16),
+    ApiInternalError(String),
+    SerializationError(serde_json::Error),
+    UnexpectedEmptyResponse { expected_type: String },
+}
+
+impl From<SiaApiClientError> for String {
+    fn from(e: SiaApiClientError) -> Self { format!("{:?}", e) }
+}
+
+/// Generic function to fetch data from a URL and deserialize it into a specified type.
+async fn fetch_and_parse<T: DeserializeOwned>(client: &Client, request: Request) -> Result<T, SiaApiClientError> {
+    let url = request.url().clone(); // TODO remove this once reqwest crate is updated
+    let fetched = client.execute(request).await.map_err(|e| {
+        SiaApiClientError::ReqwestFetchError(ReqwestErrorWithUrl {
+            error: e,
+            url: url.clone(),
+        })
+    })?;
+
+    let status = fetched.status().as_u16();
+    match status {
+        200 => {},
+        500 => {
+            // FIXME handle unwrap gracefully
+            return Err(SiaApiClientError::ApiInternalError(fetched.text().await.unwrap()));
+        },
+        _ => {
+            return Err(SiaApiClientError::UnexpectedHttpStatus(status));
+        },
+    }
+    let response_text = fetched.text().await.map_err(|e| {
+        SiaApiClientError::ReqwestParseInvalidEncodingError(
+            ReqwestErrorWithUrl {
+                error: e,
+                url: url.clone(),
+            }
+            .to_string(),
+        )
+    })?;
+
+    // Handle status 200 empty responses with marker types
+    if response_text.trim().is_empty() {
+        // Attempt to deserialize as EmptyResponse marker struct
+        if let Ok(parsed) = serde_json::from_str::<T>("null") {
+            return Ok(parsed);
+        }
+    }
```

Comment:
Any endpoint that returns HTTP 200 with an empty response is handled via `pub struct EmptyResponse;`.  I couldn't find a clean way of returning `EmptyResponse` in this case. I would need to implement an additional method or implement Default for every single endpoint otherwise.

This is a workaround that tries to parse the string "null" to the response type. This should only work with types that use `EmptyResponse`.

Definitely open to suggestions because I realize how strange this looks.

Reply:
```
    if response_text.trim().is_empty() {
        if TypeId::of::<T>() == TypeId::of::<EmptyResponse>() {
            return Ok(T::default());
        }
    }
```

I wanted to accomplish something similar to this without being forced to implement default on every single endpoint response type.